src/x/cache/lru_cache.go

```diff
@@ -0,0 +1,530 @@
+	}
+
+	// Something happened during load, but we don't want to cache this - remove the entry,
+	// tell any blocked callers they can try again, and return the error
+	entry := c.entries[key]
+	c.remove(entry)
+	close(entry.loadingCh)
+	entry.loadingCh = nil
+
+	if isUncachedError {
+		return nil, uncachedErr.Err
+	}
+
+	return nil, err
+}
+
+// updateCacheEntryWithLock updates a cache entry with a new value or cached error,
+// and marks it as the most recently accessed and most recently loaded entry
+func (c *LRU) updateCacheEntryWithLock(
+	key string, expiresAt time.Time, value interface{}, err error,
+) (interface{}, error) {
+	entry := c.entries[key]
+	if entry == nil {
+		entry = &lruCacheEntry{}
+		c.entries[key] = entry
+	}
+
+	entry.value, entry.err = value, err
+
+	// Re-adjust expiration and mark as both most recently access and most recently used
+	if expiresAt.IsZero() {
+		expiresAt = c.now().Add(c.ttl)
+	}
+
+	entry.expiresAt = expiresAt
+	entry.loadTimeElt = c.byLoadTime.PushFront(entry)
+	entry.accessTimeElt = c.byAccessTime.PushFront(entry)
+	c.metrics.entries.Update(float64(len(c.entries)))
+
+	// Tell any other callers that we're done loading
+	if entry.loadingCh != nil {
+		close(entry.loadingCh)
+		entry.loadingCh = nil
+	}
+	return value, err
+}
+
+// reserveCapacity evicts expired and least recently used entries (that aren't loading)
+// until we have at least enough space for new entries.
+// NB(mmihic): Must be called with the cache mutex locked.
+func (c *LRU) reserveCapacity(n int) error {
+	// Unconditionally evict all expired entries. Entries that are expired by
+	// reloading are not in this list, and therefore will not be evicted.
+	oldestElt := c.byLoadTime.Back()
+	for oldestElt != nil {
+		entry := oldestElt.Value.(*lruCacheEntry)
+		if entry.expiresAt.After(c.now()) {
+			break
+		}
+		c.remove(entry)
+
+		oldestElt = c.byLoadTime.Back()
+	}
+
+	// Evict any recently accessed which are not loading, until we either run out
+	// of entries to evict or we have enough entries.
+	lruElt := c.byAccessTime.Back()
+	for c.maxEntries-len(c.entries) < n && lruElt != nil {
+		c.remove(lruElt.Value.(*lruCacheEntry))
+
+		lruElt = c.byAccessTime.Back()
+	}
+
+	// If we couldn't create enough space, then there are too many entries loading and the cache is simply full
+	if c.maxEntries-len(c.entries) < n {
+		return ErrCacheFull
+	}
+
+	return nil
+}
+
+// load tries to load from the loader.
+// NB(mmihic): Must NOT be called with the cache mutex locked.
+func (c *LRU) tryLoad(
+	ctx context.Context, key string, loader LoaderWithTTLFunc,
+) (interface{}, error) {
+	// If we're limiting overall concurrency, acquire a concurrency lease
+	if c.concurrencyLeases != nil {
+		select {
+		case <-ctx.Done():
+			return c.cacheLoadComplete(key, time.Time{}, nil, UncachedError{ctx.Err()})
+		case <-c.concurrencyLeases:
+		}
+
+		defer func() { c.concurrencyLeases <- struct{}{} }()
```

Comment:
nit; maybe slightly safer as:

```suggestion
		select {
		case <-ctx.Done():
			return c.cacheLoadComplete(key, time.Time{}, nil, UncachedError{ctx.Err()})
		case <-c.concurrencyLeases:
			defer func() { c.concurrencyLeases <- struct{}{} }()
		}

```